src/Bridge/Graphql/Resolver/CollectionResolverFactory.php

```diff
@@ -0,0 +1,96 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Bridge\Graphql\Resolver;
+
+use ApiPlatform\Core\Api\IdentifiersExtractorInterface;
+use ApiPlatform\Core\DataProvider\CollectionDataProviderInterface;
+use ApiPlatform\Core\DataProvider\PaginatorInterface;
+use ApiPlatform\Core\DataProvider\SubresourceDataProviderInterface;
+use GraphQL\Type\Definition\ResolveInfo;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+
+/**
+ * Creates a function retrieving a collection to resolve a GraphQL query.
+ *
+ * @author Alan Poulain <contact@alanpoulain.eu>
+ */
+final class CollectionResolverFactory extends AbstractResolverFactory implements CollectionResolverFactoryInterface
+{
+    private $collectionDataProvider;
+    private $normalizer;
+    private $identifiersExtractor;
+    private $requestStack;
+    private $paginationEnabled;
+
+    public function __construct(CollectionDataProviderInterface $collectionDataProvider, SubresourceDataProviderInterface $subresourceDataProvider, NormalizerInterface $normalizer, IdentifiersExtractorInterface $identifiersExtractor, RequestStack $requestStack, bool $paginationEnabled)
```

Comment:
IMO the request should not be injected that deep. The low level GraphQL implementation should be agnostic of the transport layer (HTTP, AMQP...).

Can't we extract the offset and other data needed in the controller layer and pass the base64-encoded values to `createCollectionResolver`?

Reply:
This resolver does not need the request to resolve the query. It's the PaginationExtension who needs to retrieve the offset and the limit.
The resolver adds the offset / limit argument of the GraphQL query to the request attributes to be used by the PaginationExtension.